Adding java sample

scgbear opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* Adding Java sample

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]/java
git checkout [branch-name]
mvn clean package
```

* Test the code
mvn clean package runs the unit tests
```
```

## What to Check
Verify that the following are valid
* Tests pass, no failures

## Other Information
<!-- Add any other helpful information that may be needed here. -->

scgbear committed:
Adding java sample

olivertowers commented on java/README.md:
```diff
@@ -1,3 +1,11 @@
-# Azure Cosmos sample for Java for Bulk ingestion and capabilities on Graph API accounts
+# GraphBulkExecutor Java Sample
```
Let's keep the original formal title so that it matches the .NET title.

olivertowers replied:
Although we should change `Graph API` to `Gremlin API` for to match the proper API name.

olivertowers commented on java/README.md:
```diff
@@ -1,3 +1,11 @@
 
-TBD
\ No newline at end of file
+This sample uses the Java SDK 4.0 for Cosmos Db to execute bulk operations against a Cosmos Database that has been configured to use the Gremlin API. It is a port of the .Net  [GraphBulkExecutorV3](https://github.com/ealsur/GraphBulkExecutorV3).
```
We don't need to call out the original V3 sample here, as the java and dotnet implementations will serve as the spec versions for other bulk executor implementations.

kushagraThapar replied:
```suggestion
This sample uses the Java SDK v4.x.x for Cosmos Db to execute bulk operations against a Cosmos Database that has been configured to use the Gremlin API. It is a port of the .Net  [GraphBulkExecutorV3](https://github.com/ealsur/GraphBulkExecutorV3).
```

olivertowers commented on java/docs/diagrams/CLIENT_USE_CASE_1.drawio:
```diff
@@ -0,0 +1,90 @@
+<mxfile host="65bd71144e">
```
Not sure if there is built in support for draw.io but github markdown does support mermaid. You may want to see about converting your diagrams to this so it can be updated inline.

```
```mermaid
  graph LR
      A[Square Rect] -- Link text --> B((Circle))
      A --> C(Round Rect)
      B --> D{Rhombus}
      C --> D
```

```mermaid
  graph LR
      A[Square Rect] -- Link text --> B((Circle))
      A --> C(Round Rect)
      B --> D{Rhombus}
      C --> D
```

scgbear replied:
TIL about Mermaid, I'll look into that. Looks very uml friendly.

scgbear replied:
Replaced the diagrams with inline mermaid content.

olivertowers commented on java/src/main/java/com/microsoft/graph/bulk/impl/GremlinVertexSerializer.java:
```diff
@@ -0,0 +1,58 @@
+
+        jsonGenerator.writeStringField(partitionKey.getFieldName(), partitionKey.getValue());
+
+        gremlinVertex.getProperties().forEach((key, value) -> {
```
We can do this in a separate commit, but we should also support flat vertex property format as well. This is where vertex properties are written as `key : scalar_value` in json as opposed to `key : array_of_objects`.
In dotnet it is called the flag which determines this is called `isMultiValuedAndMetaPropertiesDisabled`, though I feel that this is a little obtuse and instead we should go for `enableFlatVertexPropertySchema`.

scgbear replied:
I was thinking we could do that by implementing multiple objectMapper for the different serialization approaches. Was thinking the BulkGremlinObjectMapper utility class could be extended to include a getFlatVertexObjectMapper renaming the existing getBulkGremlinObjectMapper to something like getArrayVertexObjectMapper

scgbear replied:
I created an issue for this one.

olivertowers commented on java/src/main/java/com/microsoft/graph/bulk/impl/model/GremlinVertexProperty.java:
```diff
@@ -0,0 +1,23 @@
+
+    public GremlinVertexProperty(Object value) {
+        this.value = value;
+        //TODO: Validate if this should be a deterministic UUID, and if so what values should be included?
```
Vertex property ids are best thought of as a system assigned attribute but it just assigns a random UUID each time. Bulk insert bypasses the component that would normally assign the property id, so to be consistent with the format we should duplicate that here and keep it as a hidden/non-user assigned value.

scgbear replied:
With that, I was thinking of moving it out of the class and make it part of the serializer. Should reduce some of the memory footprint in there.

olivertowers replied:
Yeah I think that's fine.

scgbear committed:
Remove connection info

olivertowers commented on java/src/test/java/com/microsoft/graph/bulk/impl/CosmosDBSQLBulkExecutorIT.java:
```diff
@@ -0,0 +1,121 @@
+//package com.microsoft.graph;
```
Remove file if this is not needed

scgbear committed:
Integration tests updated with generic sample

scgbear commented:
@ealsur Can you review this PR? We got the last bits of the integration tests converted today.

ealsur commented:
I'll defer to @kushagraThapar - I am not a Java expert 😄

kushagraThapar commented:
@scgbear - thanks for putting this PR out, I am reviewing it. Will post my comments soon!

kushagraThapar commented on java/IntelliJ_Settings.xml:
```diff
@@ -0,0 +1,7 @@
+<code_scheme name="Default" version="173">
```
Do we need to include this file?
I think we should gitignore it.

kushagraThapar commented on java/pom.xml:
```diff
@@ -0,0 +1,140 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
```suggestion
    <groupId>com.azure</groupId>
```

kushagraThapar commented on java/pom.xml:
```diff
@@ -0,0 +1,140 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
+    <artifactId>GraphBulkExecutor</artifactId>
```
```suggestion
    <artifactId>azure-cosmos-graph-bulk-executor</artifactId>
```

kushagraThapar commented on java/pom.xml:
```diff
@@ -0,0 +1,140 @@
+
+    <dependencies>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
```
If we can avoid usage of lombok, we should. Reason being it has caused some dependency issues for customers in the past. We don't want customers to fall into dependency issues when using this sample, as customers would directly copy paste this.

scgbear replied:
I'll look into removing the dependance.

scgbear replied:
Removed the dependency on lombak. Needed quite a bit of boiler plate bits added to replace the functionality.

kushagraThapar replied:
Thanks @scgbear

kushagraThapar commented on java/pom.xml:
```diff
@@ -0,0 +1,140 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-cosmos</artifactId>
+            <version>${cosmosSDKVersion}</version>
```
Instead of providing an `azure-cosmos` version, can we not always rely on the latest? So basically skipping the version here at all. Something like this - https://github.com/Azure-Samples/azure-cosmos-java-sql-api-samples/blob/main/pom.xml#L48

Otherwise we would have to push updates to this regularly, and creates overhead. 
Also, when customers copy paste, they will stick to version what we use in the samples, so best to always use latest.

kushagraThapar commented on java/pom.xml:
```diff
@@ -0,0 +1,140 @@
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
nit: new line at the end.

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/impl/BulkGremlinObjectMapper.java:
```diff
@@ -0,0 +1,29 @@
+package com.microsoft.graph.bulk.impl;
```
We don't include `microsoft` in the package structure anymore, because of the rebranding of azure. 
Please change the package structure to `com.azure. ...`

kushagraThapar replied:
It is important to add the correct license headers - since this is now an official sample. 
Can you please copy paste the header from here - https://github.com/Azure-Samples/azure-cosmos-java-sql-api-samples/blob/main/src/main/java/com/azure/cosmos/examples/common/AccountSettings.java#L1
And include it in all the java source files.

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/impl/BulkGremlinObjectMapper.java:
```diff
@@ -0,0 +1,29 @@
+     * Defines an ObjectMapper that has the Serialization Modules for both GremlinVertex and GremlinEdge classes
+     */
+    public static ObjectMapper getBulkGremlinObjectMapper() {
+        var mapper = new ObjectMapper();
```
I see we are using `var` everywhere, which was introduced in java 10, which means this sample cannot support Java 8. 

We cannot use `var` because azure-cosmos Java SDKs still support java 8. So if a cosmos Java SDK customer is using Java 8, this sample won't work for them. We should use strict types in the samples.

xinlian12 replied:
for the object mapper, maybe change to initialize only once?

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/impl/README.md:
```diff
@@ -0,0 +1,55 @@
+# GraphBulkExecutor
+
+This sample provides an example of using the Java SDK 4.0 for Cosmos Db to execute bulk operations against a Cosmos Database that has been configured to use the Gremlin API. It is a port of the .Net  [GraphBulkExecutorV3](https://github.com/ealsur/GraphBulkExecutorV3).
```
```suggestion
This sample provides an example of using the Java SDK v4.x.x for Cosmos Db to execute bulk operations against a Cosmos Database that has been configured to use the Gremlin API. It is a port of the .Net  [GraphBulkExecutorV3](https://github.com/ealsur/GraphBulkExecutorV3).
```

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/impl/model/GremlinEdge.java:
```diff
@@ -0,0 +1,51 @@
+        this.properties.putAll(properties);
+    }
+
+    //TODO: expand the validation logic to include id and label
```
Is this TODO still valid?

scgbear replied:
Yes, I'll add that in though.

scgbear replied:
Added additional logic here as well as in the Vertex and PartitionKey

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/impl/model/GremlinVertexProperty.java:
```diff
@@ -0,0 +1,23 @@
+@Data
+@EqualsAndHashCode
+public class GremlinVertexProperty {
+    //TODO: Evaluate impacts to changing this to be Serializable instead of just Object
```
We should either remove these TODOs, or implement them if we have the bandwidth :)

kushagraThapar commented on java/src/main/java/com/microsoft/graph/bulk/sample/model/PersonVertex.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.Builder;
+import lombok.Data;
+
+//TODO: Create a sample vertex to demonstrate how a client specific domain class can be decorated to produce a
```
Same here!

kushagraThapar commented on java/src/main/resources/application.properties:
```diff
@@ -0,0 +1,12 @@
+sample.sql.host={.Net SDK Url for Cosmos Db Account goes here}
```
```suggestion
sample.sql.host={Cosmos Db Account host name}
```

kushagraThapar commented on java/src/main/resources/application.properties:
```diff
@@ -0,0 +1,12 @@
+sample.sql.host={.Net SDK Url for Cosmos Db Account goes here}
+sample.sql.key={host key goes here}
```
```suggestion
sample.sql.key={Cosmos Db Account key}
```

kushagraThapar commented on java/src/test/java/com/microsoft/graph/bulk/impl/CosmosDBSQLBulkExecutorIT.java:
```diff
@@ -0,0 +1,121 @@
+//package com.microsoft.graph;
```
This file is all commented out. Do we need to keep it?

scgbear replied:
Merging error on my side. RJ replaced the tests with the UploadWithBulkLoaderIT.java class.

kushagraThapar reviewed (commented):
Amazing work @scgbear . Thanks for writing this sample. 
I have few comments, some of them are blocking, please take a look at them. 
If you would like to discuss them offline, please let me know and we can jump on a call to do so.

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/impl/BulkGremlinObjectMapper.java:
```diff
@@ -0,0 +1,29 @@
+     * Defines an ObjectMapper that has the Serialization Modules for both GremlinVertex and GremlinEdge classes
+     */
+    public static ObjectMapper getBulkGremlinObjectMapper() {
+        var mapper = new ObjectMapper();
```
Each time this method is called, a new objectMapper will be created, maybe consider to change to initialize only once?

scgbear replied:
It is only called once though, see the UploadWithBulkLoader method on the com.microsoft.graph.bulk.sample.UploadWithBulkLoader<V, E> class. 

Intent is to allow the user to opt into a different object mapper, but provide a sample if they choice to. 

Not entirly sure what an alternative approach would be, that doesn't allow the user to create more than one of these, and still allow them to provide an alternative object mapper for different serialization approaches.

scgbear committed:
Update java/README.md

Co-authored-by: Kushagra Thapar <kushuthapar@gmail.com>

scgbear committed:
Update java/pom.xml

Co-authored-by: Kushagra Thapar <kushuthapar@gmail.com>

scgbear committed:
Update java/src/main/resources/application.properties

Co-authored-by: Kushagra Thapar <kushuthapar@gmail.com>

scgbear committed:
Update java/pom.xml

Co-authored-by: Kushagra Thapar <kushuthapar@gmail.com>

scgbear committed:
Update java/src/main/resources/application.properties

Co-authored-by: Kushagra Thapar <kushuthapar@gmail.com>

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/impl/ObjectToEdge.java:
```diff
@@ -0,0 +1,281 @@
+     */
+    private static void setLabelFromGetter(Class<?> clazz, GremlinEdge results, Object from)
+            throws InvocationTargetException, IllegalAccessException {
+        for (Method method : MethodUtils.getMethodsWithAnnotation(clazz, GremlinLabelGetter.class)) {
```
do we allow > 1 methods annotated with GremlinLabelGetter ?

scgbear replied:
Currently there is nothing blocking the usage of the annotation more than once. Behavior would be to use the first instance it comes to. 

Can look into adding some logic to prevent more than one of the @GremlinId, @GremlinLable, @GremlinLabelGetter and @GremlinPartitionKey annotations. Each of them shouldn't allow for more than one.

scgbear replied:
Built the annotation validations bits and integrated into the flow right before an object is converted into either a GremlinEdge or GremlinVertex.

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/impl/ObjectToEdge.java:
```diff
@@ -0,0 +1,281 @@
+
+        results.setId((String) field.get(from));
+
+        if (results.getId() == null) {
```
should we also check empty/blank string? is empty string allowed for id?

scgbear replied:
Shouldn't be, I'll add that in there.

scgbear committed:
Implementing PR suggestions

Removing TODOs that weren't required
Removing old test class
Adding IntellijSettings file to gitignore

scgbear committed:
Merge branch 'main' of https://github.com/scgbear/azure-cosmos-graph-bulk-executor into mf-main

scgbear committed:
Revert title on README doc

with Graph changed to Gremlin

olivertowers commented:
@jayantacs Please take a look at this.

scgbear committed:
Recommended changes from PR

Remove reference to v3 of GraphBulkExecutor
Change azure cosmos version to LATEST
new line at end of pom.xml

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/sample/README.md:
```diff
@@ -0,0 +1,100 @@
+
+The /resources/application.properties file defines the data required to configure the Cosmos Db the required values are:
+
+* **sample.sql.host**: This is the value provided by the Azure Cosmos Db. Ensure you use the ".NET SDK URI" which can be located on the Overview blade of the Cosmos Db Account.
```
nit: remove ".Net SDK URI"?

scgbear replied:
That's how the value is presented to the user. Need to make sure they don't use the Gremlin Endpoint...

![image](https://user-images.githubusercontent.com/3959737/158270961-f4145371-a5f2-4f75-a1a4-b6f6ebedb325.png)

scgbear committed:
Rename packages to com.azure

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/sample/UploadWithBulkLoader.java:
```diff
@@ -0,0 +1,105 @@
+        ThroughputProperties throughputProperties = ThroughputProperties.createManualThroughput(DatabaseSettings.THROUGHPUT);
+        Mono<CosmosContainerResponse> containerIfNotExists = database.createContainerIfNotExists(containerProperties, throughputProperties);
+
+        //  Create container with 400 RU/s
```
nit: the comment may not correct. The container will be created with the configured throughput, so may not be 400RU?

xinlian12 commented on java/src/main/java/com/microsoft/graph/bulk/sample/UploadWithBulkLoader.java:
```diff
@@ -0,0 +1,105 @@
+        //  </CreateContainerIfNotExists>
+
+        //Modify existing container
+        containerProperties = cosmosContainerResponse.getProperties();
```
Q: does not seem there is any change to the container properties?

scgbear replied:
The container properties are defined on lines 68-70 and would come into play if the container already existed when line 71 is called.

scgbear committed:
Add license header

scgbear committed:
Fix missing javax from pom

Remove TODO

scgbear committed:
Remove the var's

Adding explicit type declartion to allow for Java 8 support

jayanta-mondal commented on java/README.md:
```diff
@@ -1,3 +1,14 @@
+
+There are two primary components of this sample:
+
+[GraphBulkExecutor](src/main/java/com/azure/graph/bulk/impl/): Contains the source code for the reusable code that can
```
Contains the source code of the core bulk execution functionality for Gremlin API. 


Saying this should be enough - as the 'sample' part explains what one can do with that.

jayanta-mondal commented on java/README.md:
```diff
@@ -1,3 +1,14 @@
+be harnessed within your specific domain.
+
+[Sample](src/main/java/com/azure/graph/bulk/sample/): Contains an example implementation using the above logic. It has
+some flexibility in how it is run to allow for experimentation to determine how best to use the GraphBulkExecutor.
```
Contains an example implementation using the above logic. Note that, the core GraphBulkExcutor libraries takes a set of input parameter which may need to be tuned for optimal performance for a given scenario.

jayanta-mondal commented on java/README.md:
```diff
@@ -1,3 +1,14 @@
 
-TBD
\ No newline at end of file
+This sample uses the Java SDK v4.x.x for Cosmos Db to execute bulk operations against a Cosmos Database that has been
```
high-throughput bulk operations

jayanta-mondal commented on java/README.md:
```diff
@@ -1,3 +1,14 @@
+This sample uses the Java SDK v4.x.x for Cosmos Db to execute bulk operations against a Cosmos Database that has been
+configured to use the Gremlin API.
+
+This sample is useful when hitting performance issues when uploading Vertices and Edges in bulk.
```
I think we can take this line out.

jayanta-mondal commented on java/src/main/java/com/azure/graph/bulk/impl/GremlinEdgeSerializer.java:
```diff
@@ -0,0 +1,64 @@
+        jsonGenerator.writeStringField(gremlinEdge.getPartitionKey().getFieldName(), gremlinEdge.getPartitionKey().getValue());
+        jsonGenerator.writeStringField(GremlinFieldNames.EDGE_DESTINATIONV_PARTITION, destinationVertexInfo.getPartitionKey());
+
+        jsonGenerator.writeStringField(GremlinFieldNames.EDGE_DESTINATIONV, destinationVertexInfo.getId());
```
EDGE_DESTINATIONV_ID?

jayanta-mondal commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinEdgeVertexInfo.java:
```diff
@@ -0,0 +1,99 @@
+public class GremlinEdgeVertexInfo {
+    private String id;
+    private String label;
+    private String partitionKey;
```
The partition key can be non-string. It can be a number or even a boolean. Basically any Cosmos DB primitive data type.

olivertowers replied:
Consider making this field `GremlinPartitionKey` and then make `GremlinPartitionKey` support non-string partition values

jayanta-mondal commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinPartitionKey.java:
```diff
@@ -0,0 +1,22 @@
+@AllArgsConstructor
+@EqualsAndHashCode
+public class GremlinPartitionKey {
+    private String fieldName;
```
Same - can be any primitive Cosmos DB data type.

scgbear replied:
@jayantacs or @olivertowers, I'm Writing the validation for the allowed data types. Have a question on this, is a null value allowed for the partition key?

Pretty sure I have never tried to use a nullable field as a partition key.

scgbear replied:
Found the blog post on using Undefined.value() to set an empty partitionkey value. So I added a test for the value along with the other data type validations.

jayanta-mondal commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinVertex.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    //TODO: expand the validation logic to include id and label
+    public void validate() {
```
Please complete the validation -- these are important. We don't want customers to ingest a large amount of data in a format gremlin API can't read.

jayanta-mondal commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinEdge.java:
```diff
@@ -0,0 +1,53 @@
+        this.properties.putAll(properties);
+    }
+
+    //TODO: expand the validation logic to include id and label
```
Same: Please complete the validation.

scgbear committed:
Implement remaining validation bits

GremlinEdge, GremlinVertex and GremlinPartitionKey classes validation functions implemented

scgbear committed:
PR Feedback implementation

Moved properties ID assignment to the serialization step instead of maintaining on the GremlinVertex
Added isBlank checks onto the ObjectToEdge convertion logic

scgbear committed:
Fix comment

scgbear committed:
Change wording on main readme to be more consise

olivertowers commented on java/README.md:
```diff
@@ -1,3 +1,13 @@
 
-TBD
\ No newline at end of file
+This sample uses the Java SDK v4.x.x for Cosmos Db to execute high-throughput bulk operations against a Cosmos Database
```
```suggestion
This sample uses the Java SDK v4.x.x for Azure Cosmos DB to execute high-throughput bulk operations against a Gremlin API graph.
```

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/GraphBulkExecutor.java:
```diff
@@ -0,0 +1,25 @@
+ * @param <V> Defines the class for the Vertices
+ * @param <E> Defines the class for the Edges
+ */
+public interface GraphBulkExecutor<V, E> {
```
In theory there should be not restrictions on the number of distinct POJO classes passed in right?

eg. There can be a `Employee`, `Company` vertex types passed in same Iterable.

scgbear replied:
Correct, that's pretty much an leaky abstraction. I'll make sure I can plug that hole.

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/GraphBulkExecutor.java:
```diff
@@ -0,0 +1,25 @@
+     * @param edges    Edges to write to the data store
+     * @return Results of the executions
+     */
+    Flux<CosmosBulkOperationResponse<GraphBulkExecutor<V, E>>> execute(
```
Why?

I'm not clear on why we would expect the Response to have a type reference to the caller class.

What does the do exactly?

scgbear replied:
That's where the leaky abstraction is starting at, and is pretty much just bubbling up the Java API signature.

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/impl/CosmosDBSQLBulkExecutor.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.stream.StreamSupport;
+
+@Slf4j
+public class CosmosDBSQLBulkExecutor<V, E> implements GraphBulkExecutor<V, E> {
```
I think this should be internal and there should be a public static factory method somewhere.
eg.
`GraphBulkExecutor bulkExecutor = GraphBulk.createBulkExecutor(container)`

olivertowers replied:
Can also consider a builder pattern since that is more the norm for java APIs.

scgbear replied:
went with the builder pattern, it's being used quite a bit within the solution already.

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/impl/CosmosDBSQLBulkExecutor.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.stream.StreamSupport;
+
+@Slf4j
+public class CosmosDBSQLBulkExecutor<V, E> implements GraphBulkExecutor<V, E> {
```
What is the lifetime of this instance? Can it be re-used for multiple bulk execute calls?

If the intention is for this to be used multiple times, then perhaps allowUpsert should be a parameter on execute.

scgbear replied:
The client this was used for re-used it (was a singleton bean), but applied the same policies for the entire solution. But that's a specific implementation, I like moving it to a parameter as you suggest though. Greater flexibility.

olivertowers replied:
As a design note: Long term I can see the bulk executor supporting a pattern where each Gremlin item op can be defined much like the Cosmos item ops are, eg.
```
graphBulkExecutor.excecuteOperations(Flux.just(
    GraphBulkOperations.getCreateVertexOperation(vertexA),
    GraphBulkOperations.getUpsertVertexOperation(vertexB),
    GraphBulkOperations.getCreateEdgeOperation(edgeAToB),
    GraphBulkOperations.getReplaceVertexOperation(vertexC),
    GraphBulkOperations.getDeleteVertexOperation(vertexD),
    GraphBulkOperations.getPatchVertexOperation(partialVertexE),
  ...)).blockLast();
```

Given that the initial version will only support single operation type per invocation, it might make sense to name the methods matching those op types and reserve the more general method name for when we can support heterogeneous operations in a single invocation.

For example, the methods we support today can be `executeCreateOperations(...)`, `executeUpsertOperations(...)`

scgbear replied:
Pretty impactful refactor implementing this one. I moved the call to the containers execute into the sample code and focused the functionality on creating the operations.

This solves that leaky abstraction quite well, as I didn't want to hide the results of the containers execution of the bulk operation from the client implementation. Abstracting that call was never really a huge value add of this sample. 

Since it's no longer responsible for making the execution, I renamed the class to GremlinDocumentOperationCreator. Not 100% happy with that, quite wordy. 

It does push the responsibility of combining the operations back to the client, but I think that flows well with your recommendation above.

scgbear replied:
I'll need to make a pass over the documentation to make sure the refactor is reflected correctly.

olivertowers replied:
I didn't necessarily mean for this to be done in this change, hence the 'long term'. Main concern was to keep the API open to having different invocation paths for uniform/non-uniform operation patterns.

But if you made the changes now that's great!

scgbear committed:
Update java/README.md

Co-authored-by: Oliver Towers <olivert@microsoft.com>

olivertowers commented:
@scgbear Thanks for the working on this adding this sample!
I've added comments, and will do a follow-up review after you've had time to address them.

scgbear committed:
Remove dependance on lombok.extern.slf4j.Slf4j

scgbear committed:
Remove dependance on SneakyThrows

scgbear committed:
Remove GremlinEdge's dependance on lombak

scgbear committed:
Remove GremlinEdgeVertexInfo dependance on lombak

scgbear committed:
Remove GremlinPartitionKey dependence on lombak

scgbear committed:
Remove GremlinVertex dependance on lombak

scgbear committed:
Remove lombak dependency from sample models

scgbear committed:
Remove remaining dependencies on lombak

scgbear committed:
Fix documentation of running the sample

scgbear committed:
Remove generic type definitions

scgbear committed:
Add ID to destination vertex sink constant

Provides greater clarity to the user on the intended use

scgbear committed:
Add builder pattern to the CosmosDBSQLBulkExecutor

scgbear committed:
Update partitionkey to allow for any primitive data type

not just strings

scgbear committed:
Refactor of the CosmosDBSQLBulkExecutor

Focuses the functionality on it's ability to transform from a domain object into a Gremlin DB Object suitable for the Cosmos SQL SDK

Abstraction of the call to the container doesn't create much value and requires decisions to be made about what to do with the response. Decisions better suited for the client implementation.

scgbear committed:
rename back to uploadDocuments

reads better the way it was

scgbear committed:
Replace seperate diagram files with inline mermaid

scgbear committed:
Remove todo comment from documentation

scgbear committed:
Fixing linting errors

Includes moving the tinkerpop bits used for integration testing into the test package. Hope to reduce the time for developer consumption by providing greater focus on the sample implementation

scgbear committed:
Update documentation based on last refactor

scgbear committed:
Create Vertex Annotation Validator

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/GremlinDocumentOperationCreator.java:
```diff
@@ -0,0 +1,153 @@
+     */
+    public Stream<CosmosItemOperation> getEdgeUpsertOperations(Stream<Object> edges) {
+        return edges.map(e -> {
+            GremlinEdge edge = e instanceof GremlinEdge ? (GremlinEdge) e : ObjectToEdge.toGremlinEdge(e);
```
Consider moving this into getEdgeUpsertOperation()

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/GremlinDocumentOperationCreator.java:
```diff
@@ -0,0 +1,153 @@
+        });
+    }
+
+    private CosmosItemOperation getEdgeCreateOperation(GremlinEdge edge) {
```
I think it would be good if this is public so that customers can use single op creation vs creating a Stream if they don't need it.

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinEdgeVertexInfo.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
GremlinEdgeVertexInfo isn't really complete without (id, pk, label) so if we are testing for equality it needs to include all fields.

I see that this has constructors which don't include all fields but I think this will be error prone. Can the builder instance be used will the properties are being finalized and make sure the `GremlinEdgeVertexInfo` has all required properties.

olivertowers replied:
Alternative to this is to use a validate() function and have optional equality checks on the other field (only equal if either both instances are missing the field, or if both fields have the same value for the field).

scgbear replied:
Actually the DreamWorks implementaion this sample was created illustrate why all of the values aren't always needed. 

Their implementation is/was still using tinkerpop for non-bulk related operations. The tinkerpop library doesn't need all of those values to write it's values to Cosmos (only needs the id from each of the edges). 

I was going to force the population of those values, but in many of the situations the changes required to populate (when they weren't needed) was to great. 

I prefer the alternative approach for this reason. Still allows the usage of the domain object for the tinkerpop library, without requiring values it doesn't need. But when attempting to use the same instances for bulk, when they are needed, then we can validate.

scgbear committed:
Base setup and refactor for GremlinEdge validation

scgbear committed:
revert application.properties values

scgbear committed:
Edge validation bits expanded to include edge specific logic

scgbear committed:
Integrate validation into the processing flow

scgbear committed:
Expose private single item operations

Exposing the single object conversaion functions as public
Should allow for greater reuse

scgbear committed:
Expand GremlinEdgeVertexInfo

Add validate method to class
Expand equality tests to include all fields
Ensure null equatity is validated

scgbear committed:
Allow PartitionKey value to be Undefined

scgbear commented:
@olivertowers think I've addressed all of the issues identified. If you can have another look that would be great.

Thanks!

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/model/AnnotationValidationException.java:
```diff
@@ -0,0 +1,11 @@
+    public AnnotationValidationException(Class<?> clazz, List<String> validationErrors) {
+        super(String.format("%s failed validation with the following errors %s",
+                clazz.getName(),
+                validationErrors));
```
Won't you need to use a collector to generate a single string from the List?

scgbear replied:
You're correct. This is an example of the error message from the SoManyProblemsThrowsException test in the ObjectsToEdgeTest.java class as is:

```
com.azure.graph.bulk.impl.ObjectToEdgeTest$SoManyProblems failed validation with the following errors [GremlinLabel and GremlinLabelGetter required to be present on only one field or method when there is no label set on the GremlinEdge class annotation., GremlinEdge classes can only have one field with the GremlinId annotation., GremlinEdge's require one field marked with GremlinEdgeVertex annotation and the direction DESTINATION., GremlinEdge's require one field marked with GremlinEdgeVertex annotation and the direction SOURCE., GremlinEdge classes construct the Partition Key from the source vertex and the partitionKeyName value in the GremlinEdge call annotation.
```

It works, but not very user friendly of a reading experience. I updated to use the collector.joining with a lineseperator and bit better of a format string for this...


```
com.azure.graph.bulk.impl.ObjectToEdgeTest$SoManyProblems failed validation with the following errors:

* GremlinLabel and GremlinLabelGetter required to be present on only one field or method when there is no label set on the GremlinEdge class annotation.
* GremlinEdge classes can only have one field with the GremlinId annotation.
* GremlinEdge's require one field marked with GremlinEdgeVertex annotation and the direction DESTINATION.
* GremlinEdge's require one field marked with GremlinEdgeVertex annotation and the direction SOURCE.
* GremlinEdge classes construct the Partition Key from the source vertex and the partitionKeyName value in the GremlinEdge call annotation.
```

olivertowers commented on java/src/main/java/com/azure/graph/bulk/impl/model/GremlinEdgeVertexInfo.java:
```diff
@@ -90,7 +99,17 @@ public boolean equals(Object o) {
         GremlinEdgeVertexInfo c = (GremlinEdgeVertexInfo) o;
 
-        return Objects.equals(c.getId(), this.getId());
+        if (isNotEqual(c.id, this.id)) return false;
```
Update hashCode as well ? Should be about to use Objects.hash(fieldA, fieldB, ...)

olivertowers reviewed (changes requested):
Couple more things to address.

scgbear committed:
Format exception message and update hashcode

Fixing the formatting of the AnnotationValidationException message
Fixed the Hashcode on the GremlinEdgeVertexInfo to include all of the fields

olivertowers commented:
@scgbear lgtm

scgbear commented:
Thanks @olivertowers! I assume we need another approval before it's merged into the main branch? @xinlian12, @kushagraThapar, @jayantacs are you able to have another look? 

Thanks!

kushagraThapar commented:
@scgbear - I will take another look today, thanks for resolving all the comments, appreciate it.

kushagraThapar reviewed (approved):
LGTM @scgbear , thanks for your contribution, appreciate it.

scgbear commented:
Thanks @kushagraThapar! What are the next steps to get it merged.

scgbear committed:
Add license header to new classes

kushagraThapar commented:
@scgbear - if everything looks good, I will go ahead and merge it now. Please let me know!
For any future changes, or any current changes, please feel free to create more PRs or to discuss this with the SDK team and Gremlin team.

scgbear commented:
Thanks @kushagraThapar!
